File: src/2016/day5.rs
use rayon::prelude::*;

fn main() {
    // https://adventofcode.com/2016/day/5
    println!("Day 5, 2016: How About a Nice Game of Chess?");

    let (p1, p2) = generate(get_input());

    println!("  Part 1: {}", p1);
    assert_eq!("1a3099aa", p1);

    println!("  Part 2: {}", p2);
    assert_eq!("694190cd", p2);
}

fn starts_with_00000(digest: &md5::Digest) -> bool {
    digest[0] == 0 && digest[1] == 0 && digest[2] & 0xf0 == 0
}

const CHUNK_SIZE: i32 = 200_000;

fn generate(prefix: &'static str) -> (String, String) {
    let mut index = 0;
    let mut result_1 = String::new();
    let mut result_2 = vec![' '; 8];

    while result_1.len() < 8 || result_2.contains(&' ') {
        let hashes: Vec<_> = (index..index + CHUNK_SIZE)
            .into_par_iter()
            .map(|i| md5::compute(format!("{}{}", prefix, i)))
            .filter(starts_with_00000)
            .map(|hash| format!("{:x}", hash))
            .collect();

        for hash in hashes {
            if result_1.len() < 8 {
                result_1.push(hash.chars().nth(5).unwrap());
            }

            if let Some(pos) = hash.chars().nth(5).and_then(|c| c.to_digit(10)) {
                if pos < 8 && result_2[pos as usize] == ' ' {
                    result_2[pos as usize] = hash.chars().nth(6).unwrap();
                }
            }
        }

        index += CHUNK_SIZE;
    }

    (result_1, result_2.into_iter().collect())
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2016/day5.txt").trim_end()
}

#[cfg(test)]
mod tests {
    #[test]
    fn parts() {
        let results = super::generate("abc");
        assert_eq!("18f47a30", results.0);
        assert_eq!("05ace8e3", results.1);
    }
}

File: src/2016/day5.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2016/day/5
    utils.printTitle(2016, 5, "How About a Nice Game of Chess?");

    const m = utils.Monitor.init();
    defer m.deinit();

    const input = comptime utils.trimInput(@embedFile("input"));
    const result = try generate(input);
    utils.printAnswer(1, &result.part1);
    utils.printAnswer(2, &result.part2);
}

const Result = struct {
    part1: [8]u8 = undefined,
    part2: [8]u8 = .{ ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ' },
};

fn digitToChar(value: u8) u8 {
    return std.fmt.digitToChar(value, std.fmt.Case.lower);
}

fn startsWith00000(digest: []const u8) bool {
    return digest[0] == 0 and digest[1] == 0 and digest[2] & 0xf0 == 0;
}

fn generate(input: []const u8) !Result {
    var digest: [16]u8 = undefined;
    var result = Result{};
    var count1: u8 = 0;
    var count2: u8 = 0;
    var num: u32 = 1;
    var buffer: [100]u8 = undefined;

    while (count1 < 8 or count2 < 8) : (num += 1) {
        const door_id = try std.fmt.bufPrint(&buffer, "{s}{}", .{ input, num });
        std.crypto.hash.Md5.hash(door_id, digest[0..], .{});

        if (startsWith00000(&digest)) {
            if (count1 < 8) {
                result.part1[count1] = digitToChar(digest[2] & 0x0f);
                count1 += 1;
            }

            if (count2 < 8) {
                const pos = digest[2] & 0x0f;
                if (pos < 8 and result.part2[pos] == ' ') {
                    result.part2[pos] = digitToChar((digest[3] & 0xf0) >> 4);
                    count2 += 1;
                }
            }
        }
    }

    return result;
}

test "part 1,2" {
    const example =
        \\abc
    ;
    const result = try generate(example);
    try std.testing.expectEqualStrings("18f47a30", &result.part1);
    try std.testing.expectEqualStrings("05ace8e3", &result.part2);
}
